images/jupyter-hub/utils/jupyterhub_utils/authenticator.py

```diff
@@ -36,24 +36,80 @@ def authenticate(self, handler: RequestHandler, data: Dict[str, str]) -> Any:
-            if "errorMessage" in claims:
-                app_log.error(f"Failed authentication with error: {claims['errorMessage']}")
+
+        app_log.info("claims: %s", response)
+        if "cognito:username" not in response or response.get("cognito:username") is None:
+            if "errorMessage" in response:
+                app_log.error(f"Failed authentication with error: {response['errorMessage']}")
                 return
             else:
-                app_log.error(f"Failed authentication with unknown return: {claims}")
+                app_log.error(f"Failed authentication with unknown return: {response}")
                 return
-        return cast(str, claims["cognito:username"])
+        return cast(str, response["cognito:username"])
+
+    async def refresh_user(self, user, handler=None):
+        """Refresh auth data for a given user
+
+        Allows refreshing or invalidating auth data.
+
+        Only override if your authenticator needs
+        to refresh its data about users once in a while.
+
+        .. versionadded: 1.0
+
+        Args:
+         user (User): the user to refresh
+         handler (tornado.web.RequestHandler or None): the current request handler
+        Returns:
+         auth_data (bool or dict):
+             Return **True** if auth data for the user is up-to-date
+             and no updates are required.
+
+             Return **False** if the user's auth data has expired,
+             and they should be required to login again.
+
+             Return a **dict** of auth data if some values should be updated.
+             This dict should have the same structure as that returned
+             by :meth:`.authenticate()` when it returns a dict.
+             Any fields present will refresh the value for the user.
+             Any fields not present will be left unchanged.
+             This can include updating `.admin` or `.auth_state` fields.
+        """
+
+        auth_state = yield user.get_auth_state()
+
+        response: Dict[str, Any] = boto3.client("lambda").invoke(
+            FunctionName=f"orbit-{ENV_NAME}-token-validation",
+            InvocationType="RequestResponse",
+            Payload=json.dumps({"token": auth_state['access_token']}).encode("utf-8"),
+        )
+
+        if response is None:
+            app_log.error(f"Invalid Lambda response:\n{response}")
+            return False
+
+        app_log.info("claims: %s", response)
+        if "cognito:username" not in response or response.get("cognito:username") is None:
+            if "errorMessage" in response:
+                app_log.error(f"Failed authentication with error: {response['errorMessage']}")
+                return False
+            else:
+                app_log.error(f"Failed authentication with unknown return: {response}")
+                return False
+
+        if response is None:
+            return False
+
+        return True
```

Comment:
Ok , lets assume the token is not valid,  we should be able to reroute the user into a login page.  There must be some setting that allow us to do so when we return FALSE here.

Reply:
After digging into JupyterHub's handler for `refresh_user`, there isn't specific logic as to having a redirect upon returning False.  But what happens JupyterHub then calls the authenticate method again.  We currently use that to consume the query string and validate the token from the Landing Page.  

We could check if no query string exists, then have the redirect to a page that says you need to re-login using the Landing Page.  

A more complete solution would be to add the Landing Page URL as another query string on initial authentication, then pass that back to redirect from once the token is expired.

Reply:
yes, we can use the SSM parameter '/orbit/<replace with envname>/context' with the property 'LandingPageUrl'